flow/layers/layer_state_stack.h

```diff
@@ -0,0 +1,606 @@
+
+  // Return the SkCanvas delegate if the state stack has such a delegate.
+  // The state stack will only have one of an SkCanvas, Builder, or Mutators
+  // delegate at any given time.
+  // See also |builder_delegate| and |mutators_delegate|.
+  SkCanvas* canvas_delegate() { return canvas_; }
+
+  // Return the DisplayListBuilder delegate if the state stack has such a
+  // delegate.
+  // The state stack will only have one of an SkCanvas, Builder, or Mutators
+  // delegate at any given time.
+  // See also |builder_delegate| and |mutators_delegate|.
+  DisplayListBuilder* builder_delegate() { return builder_; }
+
+  // Return the MutatorsStack delegate if the state stack has such a
+  // delegate.
+  // The state stack will only have one of an SkCanvas, Builder, or Mutators
+  // delegate at any given time.
+  // See also |builder_delegate| and |mutators_delegate|.
+  MutatorsStack* mutators_delegate() { return mutators_; }
+
+  // Clears the old delegate and sets the canvas delegate to the indicated
+  // canvas (if not nullptr). This ensures that only one delegate - either
+  // a canvas, a builder, or mutator stack - is present at any one time.
+  void set_delegate(SkCanvas* canvas);
+
+  // Clears the old delegate and sets the builder delegate to the indicated
+  // buider (if not nullptr). This ensures that only one delegate - either
+  // a canvas, a builder, or mutator stack - is present at any one time.
+  void set_delegate(DisplayListBuilder* builder);
+  void set_delegate(sk_sp<DisplayListBuilder> builder) {
+    set_delegate(builder.get());
+  }
+  void set_delegate(DisplayListCanvasRecorder& recorder) {
+    set_delegate(recorder.builder().get());
+  }
+
+  // Clears the old delegate and sets the mutators delegate to the indicated
+  // MutatorsStack (if not null). This ensures that only one delegate - either
+  // a canvas, a builder, or mutator stack - is present at any one time.
+  void set_delegate(MutatorsStack* stack);
+
+  // Overrides the initial cull rect and/or transform when it is not known at
+  // the time that the LayerStateStack is constructed. Must be called before
+  // any state has been pushed on the stack.
+  void set_initial_cull_rect(const SkRect& cull_rect);
+  void set_initial_transform(const SkMatrix& matrix);
+  void set_initial_transform(const SkM44& matrix);
+  void set_initial_state(const SkRect& cull_rect, const SkMatrix& matrix);
+  void set_initial_state(const SkRect& cull_rect, const SkM44& matrix);
+
+  class AutoRestore {
+   public:
+    ~AutoRestore();
+
+   protected:
+    LayerStateStack* layer_state_stack_;
+
+   private:
+    AutoRestore(LayerStateStack* stack);
+    friend class LayerStateStack;
+
+    const size_t stack_restore_count_;
+  };
+
+  static constexpr int CALLER_CAN_APPLY_OPACITY = 0x1;
+  static constexpr int CALLER_CAN_APPLY_COLOR_FILTER = 0x2;
+  static constexpr int CALLER_CAN_APPLY_IMAGE_FILTER = 0x4;
+  static constexpr int CALLER_CAN_APPLY_ANYTHING = 0x7;
```

Comment:
Should we make this an enum for some more type safety where these get used?

Should we bother having these public instead of just having method(s) on the class that let you set/clear them? For example, it looks like `TextureLayer` sets the OPACITY flag and clears others, whereas other layers use `|=` to add flags. It's not entirely clear to me whether that's intentional or not, whereas methods would be a bit clearer.

Reply:
They are bit fields. Enum would not be appropriate since they are combined (see the ANYTHING value). I could probably make that more obvious by having the ANYTHING be an OR(|) of the 3 values if constexpr allows me to do that.

Reply:
Changed these to kConstant naming and the Anything version is now a bit-OR.

Reply:
This part of the mechanism needs more work that goes beyond just translating existing support into a new mechanism. I would like to remove the state flags from the PrerollContext and manage the congregate values inside LSS itself, but I don't want to make that part of this initial design shift.

Some of the cases involve `flags = kCallerCan...` because in those cases the code knows that those flags are exactly what can be handled in the Paint method. Some involve `|= kCallerCan...` because those pieces of code are suggesting that someone else may have made an initial determination of what can be handled and the code in question is trying to say "whatever that other code determined, I now know enough additional information to state that the following attribute can also be handled". If I come up with an interface that manages these flags via method calls, then those method calls would have to be rich enough to describe all of these situations.

For now, I'm hoping to just barely capture existing functionality in a simple form and create a platform from which we can get more sophisticated through additional work.

Reply:
Sounds good.